common/util/tree_operations.h

```diff
@@ -0,0 +1,926 @@
+
+#include <algorithm>
+#include <cstddef>
+#include <functional>
+#include <iosfwd>  // IWYU pragma: keep
+#include <iterator>
+#include <numeric>
+#include <set>
+#include <utility>
+#include <vector>
+
+#include "common/util/logging.h"
+#include "common/util/spacer.h"
+#include "common/util/type_traits.h"
+
+namespace verible {
+
+namespace tree_operations_internal {
+
+// TODO(mglb): move (Unavailable)FeatureTraits to a more generic header, or
+// replace them with something from standard library.
+
+// Type used as a placeholder for unavailable feature in type traits.
+struct UnavailableFeatureTraits {
+  static inline constexpr bool available = false;
+};
+
+// Type used as a base for available feature in type traits.
+struct FeatureTraits {
+  static inline constexpr bool available = true;
+};
+
+// TreeNodeTraits implementation details:
+
+// Alias to Node::subnodes_type if it exists.
+// Intended for use with detected_or_t.
+template <typename Node,  //
+          typename Type_ = typename Node::subnodes_type>
+using TreeNodeSubnodesType = Type_;
+
+// Defined when `Node` contains `Children()` method which returns reference to
+// a STL-like container with nodes.
+// TODO(mglb): Add support for children containers storing Node pointers.
+template <typename Node,  //
+          typename ChildrenType_ = decltype(std::declval<Node>().Children()),
+          typename =
+              std::void_t<decltype(*std::declval<Node>().Children().begin()),
+                          decltype(std::declval<Node>().Children().end())>>
+struct TreeNodeChildrenTraits : FeatureTraits {
+  // Container type which should be used for storing arrays of detached
+  // child nodes. It must be move-assignable to, and move-constructible from,
+  // a value returned by `Children()` method.
+  using container_type = verible::remove_cvref_t<
+      detected_or_t<ChildrenType_, TreeNodeSubnodesType, Node>>;
+};
+
+// Defined when `Node` contains `Value()` method.
+template <typename Node,  //
+          typename Value_ = decltype(std::declval<Node>().Value())>
+struct TreeNodeValueTraits : FeatureTraits {
+  // Type of data returned by `Value()` method, without reference and const.
+  using type = std::remove_const_t<std::remove_reference_t<Value_>>;
+  // Reference to a type returned by `Value()` matching its constness.
+  using reference = std::add_lvalue_reference_t<Value_>;
+  // Const reference to a type returned by `Value()`
+  using const_reference = std::add_lvalue_reference_t<std::add_const_t<type>>;
+};
+
+// Defined when `Node` contains `Parent()` method which dereferences to a type
+// fulfilling NodeTraits concept.
+template <typename Node,  //
+          typename Parent_ = decltype(*std::declval<Node>().Parent()),
+          typename = std::void_t<TreeNodeChildrenTraits<Parent_>>>
+struct TreeNodeParentTraits : FeatureTraits {};
+
+// BirthRank implementation details:
+
+// BirthRank implementation supporting any container.
```

Comment:
// Requires that the tree node type provides 'Parent()'.

Reply:
I've added the comment to the public function (the cited one is internal). Should it be here too?